Testcontainer에서 postgis 확장 기능 사용 불가

## 💻 발생 환경

<!-- 버그가 발생한 환경을 작성해주세요 -->

- macOS 14.5

## 📋 버그 내용

<!-- 버그에 대한 내용을 작성해주세요 -->
<!-- 예시: 로그인 시도 시, http 400 응답과 함께 로그인이 되지 않습니다. -->

- testcontainers로 테스트 실행시, postgis 확장이 적용되지 않습니다.
- 이로 인해 DDL 쿼리가 실패합니다.

#### 😇 원래 예상한 동작

<!-- 버그가 없다고 가정했을때, 예상했던 동작을 작성해주세요 -->
<!-- 예시: 로그인 API인 /api/v1/auth/login에 GET 요청시, http 200 응답과 함께 로그인이 되어야 합니다. -->

- postgis 확장 기능을 사용하는 DDL이 정상적으로 작동합니다.

#### 😈 실제 발생 화면 또는 로그

<!-- 버그가 발생한 화면 또는 로그를 첨부해주세요 -->
<!-- 예시: 로그인 시도 시, 콘솔에 에러 로그가 출력됩니다. -->

```
org.hibernate.tool.schema.spi.CommandAcceptanceException: Error executing DDL "create table running_record (average_pace float(53) not null, calorie float(53) not null, distance float(53) not null, duration_seconds integer not null, created_at timestamp(6) with time zone, emoji_id bigint not null, end_at timestamp(6) with time zone not null, id bigint generated by default as identity, member_id bigint not null, start_at timestamp(6) with time zone not null, updated_at timestamp(6) with time zone, location varchar(255) not null, route geometry(LineString, 4326), primary key (id))" via JDBC [ERROR: type "geometry" does not exist
```

```
Caused by: org.postgresql.util.PSQLException: ERROR: type "geometry" does not exist
```

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -46,6 +46,7 @@ dependencies {
     testAnnotationProcessor("org.projectlombok:lombok:1.18.34")
 
     // testcontainers
+    testImplementation("org.springframework.boot:spring-boot-testcontainers")
     testImplementation("org.testcontainers:testcontainers:1.20.0")
     testImplementation("org.testcontainers:junit-jupiter:1.20.0")
     testImplementation("org.testcontainers:jdbc:1.20.0")
@@ -53,6 +54,7 @@ dependencies {
 
     testImplementation("org.springframework.boot:spring-boot-starter-test")
     testImplementation("org.springframework.security:spring-security-test")
+    testImplementation("org.springframework.boot:spring-boot-devtools")
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
 }
 

--- a/src/test/java/com/dnd/runus/RunUsApplicationTests.java
+++ b/src/test/java/com/dnd/runus/RunUsApplicationTests.java
@@ -1,9 +1,9 @@
 package com.dnd.runus;
 
+import com.dnd.runus.annotation.IntegrationTest;
 import org.junit.jupiter.api.Test;
-import org.springframework.boot.test.context.SpringBootTest;
 
-@SpringBootTest
+@IntegrationTest
 class RunUsApplicationTests {
     @Test
     void contextLoads() {}

--- a/src/test/java/com/dnd/runus/annotation/IntegrationTest.java
+++ b/src/test/java/com/dnd/runus/annotation/IntegrationTest.java
@@ -1,0 +1,18 @@
+package com.dnd.runus.annotation;
+
+import com.dnd.runus.config.TestcontainersConfig;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.context.annotation.Import;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Testcontainers
+@SpringBootTest
+@Import({TestcontainersConfig.class})
+public @interface IntegrationTest {}

--- a/src/test/java/com/dnd/runus/config/TestcontainersConfig.java
+++ b/src/test/java/com/dnd/runus/config/TestcontainersConfig.java
@@ -1,0 +1,19 @@
+package com.dnd.runus.config;
+
+import org.springframework.boot.devtools.restart.RestartScope;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.testcontainers.service.connection.ServiceConnection;
+import org.springframework.context.annotation.Bean;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.utility.DockerImageName;
+
+@TestConfiguration(proxyBeanMethods = false)
+public class TestcontainersConfig {
+    @Bean
+    @RestartScope
+    @ServiceConnection
+    PostgreSQLContainer<?> postgreSQLContainer() {
+        return new PostgreSQLContainer<>(
+                DockerImageName.parse("imresamu/postgis:16-3.4-alpine").asCompatibleSubstituteFor("postgres"));
+    }
+}

--- a/src/test/resources/application.yml
+++ b/src/test/resources/application.yml
@@ -8,7 +8,6 @@ spring:
 
   datasource:
     driver-class-name: org.testcontainers.jdbc.ContainerDatabaseDriver
-    url: jdbc:tc:postgresql:16.3://test
 
 app:
   api:

